modules/islandora_google_scholar/islandora_google_scholar.module

```diff
@@ -24,6 +24,8 @@ function islandora_google_scholar_create_meta_tags($object) {
   else {
     $tags = array();
     $tags['citation_author'] = array();
+    // remove breaking control characters.
+    $object['MODS']->content = preg_replace('/[\x00-\x09\x0B\x0C\x0E-\x1F\x7F]/', '', $object['MODS']->content);
```

Comment:
It's not creating a new MODS datastream, it's just re-assigning it when islandora_google_scholar_create_meta_tags is fired (every page load). This is what I got after ingesting and reloading the page several times. Are you getting something different? 

![screenshot-localhost-8000-2019 04 11-15-59-45](https://user-images.githubusercontent.com/2738244/55988848-f18baa00-5c72-11e9-819a-b1746776c056.png)

Reply:
The [`content` property of datastreams is "magic"](https://github.com/Islandora/tuque/blob/787c13b95830363ce0a2c6cd71ff5da45ccaaf27/Datastream.php#L1573-L1576), which is to say: Setting it _should_ result in another version of the datastream being created.

Indeed, it does appear to... You appear to be running that on something that would not be touched by the code changed here... a large image? The given module only runs [the given function for objects of the `ir:citationCModel` or `ir:thesisCModel` models](https://github.com/Islandora/islandora_scholar/blob/7d68a016cfe16964c1d62af68e407c66430e76f8/modules/islandora_google_scholar/islandora_google_scholar.module#L268-L282).

Reply:
Changing
`$object['MODS']->content = preg_replace('/[\x00-\x09\x0B\x0C\x0E-\x1F\x7F]/', '', $object['MODS']->content);`
to something like 
`$filtered_mods = preg_replace('/[\x00-\x09\x0B\x0C\x0E-\x1F\x7F]/', '', $object['MODS']->content);`
should be enough to get around this issue, right @adam-vessey?

Reply:
@bryjbrown: Yes, assuming of course that it's this hypothetical `$filtered_mods` variable that gets used instead of the content pulled from the MODS datastream on the next line.

Another alternative to that undertaken here which should additionally have very little chance of messing with character encodings would be to more gracefully handle the error when parsing the original content... [constructing the new `SimpleXMLElement` will throw an exception if it fails to parse](https://www.php.net/manual/en/simplexmlelement.construct.php)... we could catch and log it (watchdog(_exception), drupal_set_message, more?)... so that people would have a more direct way of discovering that their XML is invalid for whatever reason. Similarly, warning/error reporting could be taken over while constructing the object (`libxml_use_internal_errors()` and `libxml_get_errors()`, kind of thing), so we can control how/if they propagate... my concerns with the present implementation... I'm always leery of these kind of runtime filtering approaches, and how they essentially just mask underlying issues. Additionally, if the MODS _does_ happen to be parseable, there's no need to filter it... less processing is usually better, especially in code for generating displays.

... as for what to do when _handling_ an error... my first thought is just to `return FALSE;` (as we do elsewhere in this function where we have nothing to embed)...

Reply:
@adam-vessey You first comment ir:citationCModel or ir:thesisCModel model is the only one going to be impacted because this is sanitizing MODS data going through the Google Scholar submodule for Google Scholar tag display. Not other cmodels should ever see this error. There was a security patch in Islandora for forms to filter this type of stuff. This PR should only be useful for those currently in the repository (new content will get filtered during ingest). I can push a quick adjustment from that feedback. 
@bryjbrown Thanks for the suggestion. 
What do you guys think about changing the $mods line to keep it simple?
`$mods = preg_replace('/[\x00-\x09\x0B\x0C\x0E-\x1F\x7F]/', '', $object['MODS']->content);`

Reply:
@DonRichards I think you and @adam-vessey are in agreement. Only ir:citationCModels and ir:thesisCModels are affected, yet you posted a screenshot showing what appears to be a large image object. Since large images aren't affected by this function, they aren't a valid way to test whether the function creates new versions of the MODS datastream on page load or not. You'd only be able to tell by testing an ir:citationCModel and ir:thesisCModel object.

Reply:
@bryjbrown Oh man, yep I did that and yes it's creating multiple versions. Changing it to `$mods = preg_replace('/[\x00-\x09\x0B\x0C\x0E-\x1F\x7F]/', '', $object['MODS']->content);` corrects that is does filter this appropriately. @adam-vessey Would you be good with this?

Reply:
@DonRichards: It should address the issue of unintentionally creating versions; however, as [mentioned](https://github.com/Islandora/islandora_scholar/pull/319#discussion_r274688803), doesn't really facilitate the discovery of the invalid data such that it might eventually be fixed.

Looks like the same issue with parsing pops up, in the "Coins" display... and really, could pop up anywhere we attempt to parse the datastream? The issue could even go so far as Solr documents being incomplete due to exceptions in GSearch when parsing? Almost... seeming like a GIGO kind of thing... invalid data causing errors, so... really, whatever ingest mechanism allowing the invalid data should be fixed? Made to validate data, at least?... maybe a batch to fix things (though identifying what needs to be fixed could be a chore)...

In any case... threw together the exception-catching business, more as an example of what it might look like: https://github.com/Islandora/islandora_scholar/compare/7.x...adam-vessey:7.x-alternative-error-handling

Reply:
OK, I think I got it worked out. There is one place in addition I see needs attention in the islandora_xml_forms/api/XMLDocument.inc L144 `$document->loadXML(islandora_sanitize_input_for_valid_xml($xml)) :` for this to work.